service/history/queue/cross_cluster_queue_processor_base.go

```diff
@@ -0,0 +1,688 @@
+				go c.Stop()
+				break processorPumpLoop
+			}
+			ackLevel, err = c.completeAckTasks(ackLevel)
+			if err != nil {
+				c.logger.Error("failed to complete ack tasks", tag.Error(err))
+				c.metricsScope.IncCounter(metrics.TaskBatchCompleteFailure)
+			}
+			updateAckTimer.Reset(backoff.JitDuration(
+				c.options.UpdateAckInterval(),
+				c.options.UpdateAckIntervalJitterCoefficient(),
+			))
+		case <-maxPollTimer.C:
+			c.notifyAllQueueCollections()
+			if time.Now().After(lastPollTime.Add(c.options.ValidationInterval())) {
+				// No polling is happening and loop all outstanding task and validate task status
+				c.validateOutstandingTasks()
+				lastPollTime = time.Now()
+			}
+			maxPollTimer.Reset(backoff.JitDuration(
+				c.options.MaxPollInterval(),
+				c.options.MaxPollIntervalJitterCoefficient(),
+			))
+		case <-c.processCh:
+			maxRedispatchQueueSize := c.options.MaxRedispatchQueueSize()
+			if c.redispatcher.Size() > maxRedispatchQueueSize {
+				// has too many pending tasks in re-dispatch queue, block loading tasks from persistence
+				c.redispatcher.Redispatch(maxRedispatchQueueSize)
+				if c.redispatcher.Size() > maxRedispatchQueueSize {
+					// if redispatcher still has a large number of tasks
+					// this only happens when system is under very high load
+					// we should backoff here instead of keeping submitting tasks to task processor
+					c.resetBackoffTimer()
+				}
+				// re-enqueue the event to see if we need keep re-dispatching or load new tasks from persistence
+				select {
+				case c.processCh <- struct{}{}:
+				default:
+				}
+				continue processorPumpLoop
+			}
+			c.processQueueCollections()
+		case notification := <-c.actionNotifyCh:
+			c.handleActionNotification(notification)
+			lastPollTime = time.Now()
+		}
+	}
+}
+func (c *crossClusterQueueProcessorBase) completeAckTasks(ackLevel int64) (int64, error) {
+	notification := actionNotification{
+		action:               NewGetStateAction(),
+		resultNotificationCh: make(chan actionResultNotification, 1),
+	}
+	c.handleActionNotification(notification)
+	var newAckLevel task.Key
+	select {
+	case resultNotification := <-notification.resultNotificationCh:
+		if resultNotification.err != nil {
+			return ackLevel, resultNotification.err
+		}
+		for _, queueState := range resultNotification.result.GetStateActionResult.States {
+			newAckLevel = minTaskKey(newAckLevel, queueState.AckLevel())
+		}
+	case <-c.shutdownCh:
+		return ackLevel, nil
+	}
+	newAckLevelTaskID := newAckLevel.(transferTaskKey).taskID
+	c.logger.Debug(fmt.Sprintf("Start completing cross cluster task from: %v, to %v.", ackLevel, newAckLevelTaskID))
+	c.metricsScope.IncCounter(metrics.TaskBatchCompleteCounter)
+
+	if ackLevel >= newAckLevelTaskID {
+		return ackLevel, nil
+	}
+
+	if err := c.shard.GetExecutionManager().RangeCompleteCrossClusterTask(context.Background(), &persistence.RangeCompleteCrossClusterTaskRequest{
```

Comment:
nit: probably use a finite and small timeout for this persistence op so that it won't hang or block the entire processing loop for too long?

Reply:
updated